lib/puppet/functions/inverted_hashlist.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+# Copyright (c) 2018 The Regents of the University of Michigan.
+# All Rights Reserved. Licensed according to the terms of the Revised
+# BSD License. See LICENSE.txt for details.
+
+Puppet::Functions.create_function(:inverted_hashlist) do
+  dispatch :from_lookup do
+    required_param 'String', :lookup_string
+    return_type 'Hash[String, Array[String]]'
+  end
+
+  dispatch :from_hash do
+    required_param 'Hash[String, Array[String]]', :input_hash
+    return_type 'Hash[String, Array[String]]'
+  end
+
+  def from_lookup(lookup_string)
+    from_hash(call_function('lookup', lookup_string))
+  end
+
+  def from_hash(input_hash)
+    Hash.new { |h, k| h[k] = [] }.tap do |result|
+      input_hash.each do |key, values|
+        values.each do |value|
+          result[value] << key
+          result[value].sort!
```

Comment:
I'm not sure how better to test that `{a: [1], b: [1]}` returns `{1: [a, b]}`, since it might also return `{1: [b, a]}` and still be what we want but fail the test. I couldn't find an easy way to assert, say, that `result['1']` has length 2 and contains both `a` and `b` or that the resulting array is either `[a, b]` or `[b, a]`.

All I have to work with without doing anything even hackier than sorting each array is `.and_returns()`, which doesn't give me a lot of room. Any ideas?